returnn/tf/engine.py

```diff
@@ -2316,27 +2320,76 @@ def extra_fetches_callback(seq_idx, seq_tag, **kwargs):
 
       outputs, beam_scores, targets = [], [], []
       # noinspection PyShadowingNames
-      for target_idx in range(num_targets):
-        outputs.append(kwargs["output_" + output_layer_names[target_idx]])
-        beam_scores.append(kwargs["beam_scores_" + output_layer_names[target_idx]])
-        if do_eval:
-          targets.append(kwargs["target_" + target_keys[target_idx]])
+      for output_layer_idx in range(num_output_layers):
+        outputs.append(kwargs["output_" + output_layer_names[output_layer_idx]])
+        beam_scores.append(kwargs["beam_scores_" + output_layer_names[output_layer_idx]])
+        if do_eval and target_keys[output_layer_idx] is not None:
+          targets.append(kwargs["target_" + target_keys[output_layer_idx]])
+        else:
+          targets.append(None)
 
       n_batch = len(seq_idx)  # without beam
       assert n_batch == len(seq_tag)
 
       # noinspection PyShadowingNames
-      for target_idx in range(num_targets):
-        if beam_scores[target_idx] is not None:
-          assert beam_scores[target_idx].shape == (n_batch, out_beam_sizes[target_idx])
+      for output_layer_idx in range(num_output_layers):
+        if beam_scores[output_layer_idx] is not None:
+          assert beam_scores[output_layer_idx].shape == (n_batch, out_beam_sizes[output_layer_idx])
 
-        assert n_batch * (out_beam_sizes[target_idx] or 1) == len(outputs[target_idx])
-        if do_eval and targets[target_idx] is not None:
-          assert n_batch == len(targets[target_idx])
+        assert n_batch * (out_beam_sizes[output_layer_idx] or 1) == len(outputs[output_layer_idx])
+        if do_eval and targets[output_layer_idx] is not None:
+          assert n_batch == len(targets[output_layer_idx])
 
-        if output_layers[target_idx].output.dim == 256 and output_layers[target_idx].output.sparse:
+        if output_layers[output_layer_idx].output.dim == 256 and output_layers[output_layer_idx].output.sparse:
           # Interpret output as bytes/utf8-string.
-          outputs[target_idx] = bytearray(outputs[target_idx]).decode("utf8")
+          outputs[output_layer_idx] = bytearray(outputs[output_layer_idx]).decode("utf8")
+
+      serialized_outputs = []
+      serialized_targets = []
```

Comment:
Type of these two lists should be defined, like above.
Also, some comment should be added on the expected length. I assume also `num_targets`?

Reply:
Or rather `num_output_layers` now.